Update Use OpenID Connect to Build a Simple Node.js Website (#1043)
oktadeveloper/okta-blog: _source/_posts/2017-10-19-use-openid-connect-to-build-a-simple-node-website.md

tags: [javascript, oidc, openid, nodejs, expressjs]
tweets:
    - "Learn how to use OpenID Connect to build a simple Node website 🛠"
    - "Come see how to build a simple Node website using OpenID Connect!"
type: conversion
image: blog/use-openid-connect-to-build-a-simple-node-website/openid-node-social.jpg
github: https://github.com/oktadev/okta-express-example
changelog:
  - 2022-01-20: Updated dependencies, code, and the cover image. See this post's changes in [okta-blog#1043](https://github.com/oktadev/okta-blog/pull/1043) and the example app changes in [okta-express-example#3](https://github.com/oktadev/okta-express-example/pull/3).
---

If you've ever spent time trying to figure out the best way to handle user
authentication for your Node app and been confused: you're not alone. Over the
last few years, authentication practices have changed quite a bit.

Today I'm going to show you how to use OpenID Connect to build an extremely
simple Node.js website (using Express.js) that allows you to manage your users,
log them in, and log them out.

> All websites would require users to register with a

Reviewer: ```suggestion
Websites used to require users to register with a
```
the current construction is over-complicated and a bit awkward...Also, I would avoid saying  "All"

> username/password and log in with those same credentials back in the day. This method was simple but

Reviewer: ```suggestion
username/password and log in with those same credentials. This method was simple but
```
"back in the day" seems to dangle as a closing phrase, but I don't think it's necessary here.

caused many security problems because developers would need to write the
code to authenticate the user directly, store their credentials, manage their

> data, etc. It also required developers to build custom authorization schemes to track what permissions their users had to perform certain

Reviewer: ```suggestion
data, etc. This practice also required developers to build custom authorization schemes to track the user permissions necessary to perform certain 
```
simplified

operations.

A while later, OAuth came into fashion with a new idea: let a user have one
account with a large OAuth provider (Google, Facebook, etc.) and allow users to log in to your service via their OAuth account with that provider. This method had some excellent
benefits: developers no longer needed to worry about storing passwords and managing
credentials. The downside was that OAuth is a flexible protocol and doesn't lay
out rules around authorization, data management, etc. This means that developers
using pure OAuth must write a lot of custom security code themselves,
which causes problems.

> Then, [OpenID Connect](http://openid.net/connect/) (OIDC) has

Reviewer: ```suggestion
Then, [OpenID Connect](http://openid.net/connect/) (OIDC) 
```
verb tense

> come onto the scene. It's a protocol built on top of OAuth that provides

Reviewer: ```suggestion
came onto the scene. It's a protocol built on top of OAuth that provides
```

- `cookie.httponly`: this option tells the browser that JavaScript code should
  not be allowed to access the session data. JavaScript on clients is a

>   dangerous thing. Ensuring your cookies that contain identity information are

Reviewer: ```suggestion
  dangerous thing. It's essential to ensure the safety of cookies that contain identity information.
```
this sentence is super convoluted, if you're going to edit, let's remove passive voice and awkward syntax.

>   safe is always of utmost importance.

Reviewer: ```suggestion

```
see above

more about OIDC. You can [check it out
here](/blog/2017/07/25/oidc-primer-part-1).

> You can also [follow me](https://twitter.com/rdegges) and

Reviewer: ```suggestion
You can also [follow OktaDev](https://twitter.com/oktadev) and
```
do we need to keep linking to an advocate who's no longer working at Okta? i'm not sure I would... but feel free to disregard this comment...

> [Okta](https://twitter.com/oktadev) on Twitter to see more of what I'm working

Reviewer: ```suggestion
[Okta](https://twitter.com/oktadev) on Twitter to see more of what we're working on at Okta or follow [Randall](https://twitter.com/rdegges) for my latest work. 
```
here's an equitable workaround
